src/JsonSchema/SchemaFactory.php

```diff
@@ -258,19 +258,19 @@ private function getMetadata(string $className, string $type = Schema::TYPE_OUTP
 
         return [
             $resourceMetadata,
-            $serializerContext ?? $this->getSerializerContext($resourceMetadata, $type, $operationType, $operationName),
+            $this->getSerializerContext($resourceMetadata, $type, $operationType, $operationName, $serializerContext),
             $inputOrOutput['class'],
         ];
     }
 
-    private function getSerializerContext(ResourceMetadata $resourceMetadata, string $type = Schema::TYPE_OUTPUT, ?string $operationType, ?string $operationName): array
+    private function getSerializerContext(ResourceMetadata $resourceMetadata, string $type = Schema::TYPE_OUTPUT, ?string $operationType, ?string $operationName, ?array $previousSerializerContext): array
     {
         $attribute = Schema::TYPE_OUTPUT === $type ? 'normalization_context' : 'denormalization_context';
 
         if (null === $operationType || null === $operationName) {
-            return $resourceMetadata->getAttribute($attribute, []);
+            return array_merge($resourceMetadata->getAttribute($attribute, []), (array) $previousSerializerContext);
         }
 
-        return $resourceMetadata->getTypedOperationAttribute($operationType, $operationName, $attribute, [], true);
+        return array_merge($resourceMetadata->getTypedOperationAttribute($operationType, $operationName, $attribute, [], true), (array) $previousSerializerContext);
```

Comment:
Is the intention only to preserve `CONTEXT_SERIALIZATION_FORMAT_OPENAPI_PRE_V3_0`? If so, perhaps we'll have less complications just explicitly doing that outside of this function.

Reply:
This code generates a "default serializer context": if we push the context out of here, we add conditionals around all of it when only that one key is defined.

Instead, here we make it less messy, by always producing an `array` (be it already filled or not), and having all code around this detail ignore the specifics.

After all, this function is specifically called `getSerializerContext`: it is its job to prepare the serializer context for operation.

Reply:
Hmm... Indeed it's not the previous serializer context, but the provided serializer context, right?

Before: Use the provided serializer context, otherwise get the default from the resource metadata.
After: Always get the serializer context from the resource metadata, but merge it into the provided serializer context.

This change seems wrong to me, as it removes the possibility of overriding the serializer context (instead of getting it from the resource metadata), which looks to me to be the intention of the original design.

Reply:
The original code provided its own serializer context *if* none was given.

Note that the added flag is not supposed to be given from the outside: we already have all the "schema version" flags for that.

Reply:
Consider this:

```php
$schemaFactory->buildSchema(..., $myCustomSerializerContext, ...)
```

Before this change, the provided serializer context would get used, overriding the serializer context defined in the resource metadata.

But after this change, that's no longer the case, as the provided serializer context merely gets merged into the serializer context defined in the resource metadata.

Reply:
No, that's exactly why the `$previousSerializerContext` is the last argument here: its values need to be preserved.

Reply:
I think we're talking past each other here. Sorry... Let me try to rephrase:

The provided serializer context is supposed to override the serializer context defined in the resource metadata (which is used by default, if no serializer context is passed in).

Reply:
Sure, the `array_merge` keeps the keys from the passed in serializer context, but it'd still be a problem if the serializer context from the resource metadata has additional keys.

Reply:
From what I saw thus far, it only did so when none was provided.

Do you have a test case that fails with this diff, and passes with the `2.5` state? I think that would highlight the issue in more detail.

Reply:
> it'd still be a problem if the serializer context from the resource metadata has additional keys.

Does it make sense to override keys that are user-defined here (at all)? What would the keys be, for example?

Reply:
Example:

Serializer context in resource metadata:

```php
[
    'a' => 1,
    'b' => 2,
]
```

Serializer context passed in to `SchemaFactoryInterface::buildSchema`:

```php
[
    'a' => 1,
]
```

Expected (before this change) - passed in serializer context overrides what's defined in the resource metadata:

```php
[
    'a' => 1,
]
```

Actual (after this change) - passed in serializer context gets merged into what's defined in resource metadata:

```php
[
    'a' => 1,
    'b' => 2,
]
```

Reply:
Hmm, at this point I won't get to fix (or inspect) that, because I'm off for a few weeks starting tomorrow: would you be OK if we documented this into a new issue?

Reply:
BTW, what kind of issue could arise by inheriting the serializer context, instead of explicitly overriding it?

Reply:
About the potential impact, I don't know. There are so many things that could be in the serializer context (from Symfony itself, from API Platform, and of course any custom keys from the user).

Reply:
Yep, `array` is indeed a squishy API. Will investigate then :)

Reply:
Resolved in https://github.com/Ocramius/core/commit/9d17089d1c75d65122bdd092c25dce00d2bf0763

Reply:
Resolved in https://github.com/Ocramius/core/commit/9d17089d1c75d65122bdd092c25dce00d2bf0763